Implement traefik as ddev-router, fixes #3458, fixes #2653 (#4248)
drud/ddev: docs/content/users/configuration/experimental.md

* Rancher Desktop does not provide bind mounts, so use `ddev config global --no-bind-mounts` which also turns on Mutagen.
* Use a non-`ddev.site` name, `ddev config --additional-fqdns=rancher` for example, because the resolution of `*.ddev.site` seems to make it not work.
* Rancher Desktop does not seem to currently work with `mkcert` and `https`, so turn those off with `mkcert -uninstall && rm -r "$(mkcert -CAROOT)"`. This does no harm and can be undone with just `mkcert -install` again.

## Traefik Router

> The job of the `ddev-router` is to receive most HTTP or HTTPS traffic, like requests to `*.ddev.site`, and deliver it to the correct project's web container, see [Container Architecture](../basics/architecture.md#container-architecture). The router as of DDEV v1.21.3 was always a forked, poorly documented nginx reverse proxy container.

Reviewer: Edit for brevity:

```suggestion
DDEV’s router plays an important role in its [container architecture](../basics/architecture.md#container-architecture), receiving most HTTP and HTTPS traffic for requests like `*.ddev.site` and delivering it to the relevant project’s web container.
```

> The very popular [Traefik Proxy](https://traefik.io/traefik/) will replace the current `ddev-router` in the future, but is available as an experimental configuration today. Feedback is welcome. To enable it:

Reviewer: Combine explanation of history and availability of replacement, a bit more concisely:

```suggestion
Since v1.21.3, `ddev-router` has been based on a forked, poorly-documented nginx reverse proxy. A future release will replace this with the popular [Traefik Proxy](https://traefik.io/traefik/), available now as an experimental feature. Run the following to enable it:
```

> The very popular [Traefik Proxy](https://traefik.io/traefik/) will replace the current `ddev-router` in the future, but is available as an experimental configuration today. Feedback is welcome. To enable it:

Author: ```suggestion
Until DDEV v1.21.3, `ddev-router` was been based on a forked, poorly-documented nginx reverse proxy. A future release will replace this with the popular [Traefik Proxy](https://traefik.io/traefik/), available now as an experimental feature. Run the following to enable it:
```

Reviewer: Whoops! I misread that as `ddev-router` starting with its current forked nginx proxy in v1.21.3.

Your edit fixes my mistake, but also seems to suggest that in DDEV v1.21.3 `ddev-router` is no longer based on that nginx fork—but it still is by default if I’m understanding correctly. I edited my suggestion above to clarify that 1.21.3 adds an opt-in Traefik option that will become the default in some future release:

> The default `ddev-router` is based on a forked, poorly-documented nginx reverse proxy. A future release will replace this with the popular [Traefik Proxy](https://traefik.io/traefik/), available in DDEV v1.21.3 as an experimental feature. Run the following to enable it:

Reviewer: ...and one more revision to try and clarify that 1.21.3 adds a new option for the first time. It’s weird to start the paragraph calling it a “default” when it’s been the _only_ option before this new introduction:

> `ddev-router` has been based on a forked, poorly-documented nginx reverse proxy. DDEV v1.21.3 adds a new router based on the popular [Traefik Proxy](https://traefik.io/traefik/), available as an experimental feature until it becomes the default in a future release. Run the following to enable it:

Updated my suggestion above and I promise I’ll leave it alone now.

```
ddev poweroff && ddev config global --use-traefik
```

> Most DDEV projects will work fine out of the box, but there is a vast array of possible configuration changes and options, and new ways to view what is going on with the router. You don't need to know any of these to use the Traefik router, but they will offer new options in the future.

Reviewer: Edit for brevity, move “feedback welcome” sentiment alongside future feature note:

```suggestion
Most DDEV projects will work fine out of the box, with the benefit of vastly more configuration options and ways to work with the router. (This will likely lead to more features in the future, and we’d love your feedback if you’re trying this out now!)
```

### Traefik Configuration

> All configuration for the new router is intended to be customizable where needed. As with other files throughout the DDEV ecosystem, if the file has `#ddev-generated` in it, it can and will be overwritten by DDEV. If you want to "take over" the configuration, you remove the `#ddev-generated` and become responsible for the file's content.

Reviewer: Shorten and strengthen start, link immediately to configuration docs:

```suggestion
You can fully customize the router’s [Traefik configuration](https://doc.traefik.io/traefik/).
```

> All Traefik configuration is described at [docs.traefik.io](https://doc.traefik.io/traefik/).

Reviewer: Mention the next-most-important and Traefik-specific provider type next:

```suggestion
All Traefik configuration uses the *file* provider, not the *docker* provider. Even though the Traefik daemon itself is running inside the `ddev-router` container, it uses mounted files for configuration, rather than listening to the Docker socket.
```

> All Traefik configuration uses the *file* provider, not the *docker* provider. Even though the Traefik daemon itself is running inside the `ddev-router` container, it uses mounted files for configuration, rather than listening to the Docker socket.

Reviewer: End on config takeover as an aside, since that’s _not_ Traefik-specific and it may already be a familiar pattern:

```suggestion
!!!tip
    Like other DDEV configuration, any file with `#ddev-generated` will be overwritten unless you choose to “take over” it yourself. You can do this by removing the `#ddev-generated` line. DDEV will stop making changes to that file and you’ll be responsible for updating it.
```

#### Global Traefik Configuration

> Global configuration for Traefik is automatically generated in ~/.ddev/traefik.

Reviewer: Use backticks and “directory”:

```suggestion
Global configuration is automatically generated in the `~/.ddev/traefik` directory:
```

(Also removes redundant “Traefik” here since it’s in the heading and the directory name and hardly ambiguous at this point.)

> * `static_config.yaml` which is the base configuration.

Reviewer: Keep format consistent:

```suggestion
* `static_config.yaml` is the base configuration.
```

> * `certs/default_cert.*` is the default DDEV-generated certificates.

Reviewer: Fix singular+plural conundrum (_is the (...) certificates_):

```suggestion
* `certs/default_cert.*` files are the default DDEV-generated certificates.
```

* `config/default_config.yaml` contains global dynamic configuration, including pointers to the default certificates.

#### Project Traefik Configuration

> Project configuration is automatically generated in the project's .ddev/traefik directory.

Reviewer: Backticks and curly apostrophe:

```suggestion
Project configuration is automatically generated in the project’s `.ddev/traefik` directory.
```

Author: All of which I fundamentally object to, as you know :)  We'll have to find some truce on that. Probably figuring out a way to do *presentation* with curlies without affecting the *text*.

Reviewer: Wait—I know you don’t like curly quotes for text, but you don’t object to the _backticks_ do you?

Author: No, of course can't ever object to backticks. Skimming thought it had to do with curly quotes.

> * The `certs` directory contains the `<projectname>.crt` and `<projectname>.key` certificate generated for this project.

Reviewer: Swap “this” for “the” to be consistent. And because we’re talking about a per-project file and not a specific project:

```suggestion
* The `certs` directory contains the `<projectname>.crt` and `<projectname>.key` certificate generated for the project.
```

* The `config/<projectname>.yaml` file contains the configuration for the project, including information about routers, services, and certificates.

### Debugging Traefik Routing

> * Traefik provides a dynamic description of its configuration that you can visit at `http://localhost:9999`.

Reviewer: Un-bullet since there’s no introduction or strong reason for this to be a list item:

```suggestion
Traefik provides a dynamic description of its configuration you can visit at `http://localhost:9999`.
```

> * When things seem to be going wrong, do a `ddev poweroff` and then start your project and see what the Traefik daemon is doing or failing at with `docker logs ddev-router` or `docker logs -f ddev-router`.

Reviewer: Un-bullet for the same reason, “run” command rather than “do”, link to command reference, and attempt to subtly clarify the relationship between the container logs and daemon we’re looking to investigate:

```suggestion
When things seem to be going wrong, run [`ddev poweroff`](../basics/commands.md#poweroff) and then start your project again by running [`ddev start`](../basics/commands.md#start). Examine the router’s logs to see what the Traefik daemon is doing (or failing at) by running `docker logs ddev-router` or `docker logs -f ddev-router`.
```
